runtime/lua/vim/filetype/detect.lua

```diff
@@ -0,0 +1,456 @@
+local M = {}
+
+---@private
+local function count_lines(bufnr)
+  return vim.api.nvim_buf_line_count(bufnr)
+end
+
+---@private
+local function get_lines(bufnr, start_lnum, end_lnum)
+  end_lnum = end_lnum or start_lnum
+  return vim.api.nvim_buf_get_lines(bufnr, start_lnum - 1, end_lnum, false)
+end
+
+---@private
+local function concat_lines(bufnr, start_lnum, end_lnum)
```

Comment:
I think we should combine these two functions by adding an `opts` table with a `concat` key (default whatever is simpler).

If default `false`, we can still special case if no `end_lnum` and `opts` is passed, in which case it just returns `nvim_buf_get_lines(bufnr, start_lnum-1, start_lnum, false)[1]` -- that should simplify the calling code as well.